language/compiler/ir-to-bytecode/syntax/src/syntax.rs

```diff
@@ -1048,81 +1087,98 @@ fn parse_statement(tokens: &mut Lexer) -> Result<Statement, ParseError<Loc, anyh
+    tokens: &mut Lexer,
+) -> Result<Vec<(BlockLabel, Block_)>, ParseError<Loc, anyhow::Error>> {
+    let entry_loc = tokens.previous_end_loc();
+    let mut labels = HashSet::new();
+    let mut blocks = vec![];
+    let mut current_label = None;
+    let mut statements = vec![];
     while tokens.peek() != Tok::RBrace {
-        stmts.push(parse_statement(tokens)?);
+        let statement = parse_statement(tokens)?;
+        match statement {
+            Statement::LabelStatement(label) => {
+                if labels.contains(&label) {
+                    return Err(ParseError::InvalidToken {
+                        location: label.loc,
+                        message: format!("re-declaration of block label {}", label.value),
+                    });
+                }
+                labels.insert(label.clone());
+                blocks.push((
+                    match current_label {
+                        Some(label) => label,
+                        None => spanned(
+                            tokens.file_name(),
+                            entry_loc,
+                            entry_loc,
+                            BlockLabel_(Symbol::from("<entry>")),
+                        ),
+                    },
+                    Block_::new(statements),
+                ));
+                statements = vec![];
+                current_label = Some(label);
+            }
+            _ => statements.push(statement),
+        }
+    }
+    blocks.push((
+        match current_label {
+            Some(label) => label,
+            None => spanned(
+                tokens.file_name(),
+                entry_loc,
+                entry_loc,
+                BlockLabel_(Symbol::from("<entry>")),
+            ),
+        },
+        Block_::new(statements),
+    ));
+
+    // Verify all `if` and `goto` statements' labels actually exist.
+    let undeclared_label_error = |label: &BlockLabel| {
+        Err(ParseError::InvalidToken {
+            location: label.loc,
+            message: format!("use of undeclared block label {}", label.value),
+        })
+    };
+    for statement in blocks.iter().map(|b| &b.1.statements).flatten() {
+        if let Statement::CommandStatement(command) = statement {
+            match &command.value {
+                Cmd_::IfElse(_, true_label, false_label) => {
+                    if let Some(label) = true_label {
+                        if !labels.contains(label) {
+                            return undeclared_label_error(label);
+                        }
+                    }
+                    if !labels.contains(false_label) {
+                        return undeclared_label_error(false_label);
+                    }
+                }
+                Cmd_::Goto(label) => {
+                    if !labels.contains(label) {
+                        return undeclared_label_error(label);
+                    }
+                }
+                _ => {}
+            }
+        }
```

Comment:
Yeah, I winced at this part a little while I was writing it. I consulted with someone who'd written a "single-pass C compiler," as they described it, and I asked them "how could you possibly implement `goto` and label statements in a single pass?" It turns out their answer was that they didn't, and from what I understand, I believe it's impossible to implement jumps and labels with a single pass.

> I wonder if we should do it in the parser as it makes parsing no longer context free (or at least less so). Does this seem worrying to you guys?

I'm not super pleased with it, so I'm very open to alternative suggestions. By this statement do you mean that you think it'd be better to put this check in a separate part of the codebase? Or that we can avoid this check entirely by implementing this feature in a different way?

Reply:
Yeah I kind of skipped this section as I know I can tend to be a bit over-opinionated on how code is structured and have been  trying to give people some more freedom on PRs. But since y'all asked.... let me sketch something up

Reply:
Speaking of missing comments, I wrote something out explaining how this is already solved in other parts of the compiler and I now can't find that comment.

Maybe we can just talk 1:1 about it

Reply:
Just for posterity, Todd and I discussed this 1:1 via direct messages, and I'll be borrowing inspiration from how the Move language compiler does something similar, for example in its function `compile_function_body_bytecode`.

Reply:
Ah, OK, I think I understand. I got a little confused by the conversation here: I believe @vgao1996's point is that I should parse Move IR in a context-free way. Here, I think that means that parsing should be able to complete successfully regardless of whether labels are defined or not. That, I can do!

I also think that users ought to be informed that they have made an error should they write Move IR such as `goto my_label;`, but without defining `my_label:` anywhere. To do so, I think I'll need to analyze the parsed program. The Move IR `compiler` crate currently does not perform any analysis after parsing and before lowering to bytecode, so I'll add a module to perform the analysis. Easy peasy!

I got confused when I conflated "context-free parsing" with "do this in a single pass." @tnowacki's suggestion to solve this by reusing the fake-to-actual label mappings utilized in other parts of the compiler is, I think, meant to provide one way to implement branch labels, but if I add code like that to `language/compiler/ir-to-bytecode/syntax/src/syntax.rs`, then it'll still be an implementation of a contextual grammar, and so won't address the problem here.

(Meanwhile, looking for myself, it looks like the fake-to-actual mapping strategy is used by the ir-to-bytecode crate to translate `move_ir_types::ast::{Branch, BrTrue, BrFalse}` enum variants, which use strings as branch destination values, to `move_binary_format::file_format::{Branch, BrTrue, BrFalse}` enum variants, which use `u16` as branch destination values -- these represent the offset of the destination instruction from the beginning of the function it belongs to. When performing the translation, it's assumed that all string labels referenced by `ast::Branch` are also included in the vector of string labels passed in to the ir-to-bytecode crate. If not, the crate's `remap_branch_offsets` function panics -- that function iterates over every bytecode instruction, checks if it is `Branch`, `BrTrue`, or `BrFalse`, and if so, performs an unchecked access into a map from "fake" offset to actual offset. So, I'm not sure this strategy will help me here: I need to check that string labels exist, not map string labels to integer offsets.)

Reply:
I moved this logic to a separate verification routine in the ir-to-bytecode crate, and now the IR compiler calls that to check that all labels are declared only once, and no undeclared labels are used.